src/station_cmd.cpp

```diff
@@ -2449,6 +2449,53 @@ bool HasStationInUse(StationID station, bool include_company, CompanyID company)
 	return false;
 }
 
+/**
+ * Ensure this tile is not a target for ships to get to a nearby dock.
+ * @note Also takes into account Oil Rigs, even those under construction.
+ * @param tile Tile to query
+ * @param diagdir Optional diagonal direction to look for axis compatibility.
+ * @return Succeeded if the tile is not a ship docking area of a nearby dock,
+ *         or an error message if the tile is a ship docking area of a nearby
+ *         dock.
+ * @note If the axis of a provided diagdir is equal to the axis of a nearby
+ *       docking area, it returns Succeeded. If diagdir is not provided, only
+ *       the docking area location is ensured.
+ */
+static CommandCost EnsureNoDockingTile(TileIndex tile, DiagDirection diagdir = INVALID_DIAGDIR)
+{
+	Axis axis = diagdir == INVALID_DIAGDIR ? INVALID_AXIS : DiagDirToAxis(diagdir);
+
+	for (DiagDirection rot = DIAGDIR_BEGIN; rot < DIAGDIR_END; rot++) {
+		TileIndex tc = TileAddByDiagDir(tile, rot);
+		TileIndexDiff diff = TileDiffXY(-1, 0);
+		TileIndex td;
+		if ((IsTileType(tc, MP_STATION) && IsDock(tc) && (td = tc)) || (IsTileType(tc + diff, MP_STATION) && IsOilRig(tc + diff) && (td = tc + diff))) {
+			Station *dock = Station::GetByTile(td);
+			TileIndex docking_location = TILE_ADD(dock->dock_tile, ToTileIndexDiff(GetDockOffset(dock->dock_tile)));
+			if (axis == INVALID_AXIS) {
+				return_cmd_error(STR_ERROR_TOO_CLOSE_TO_ANOTHER_DOCK);
+			} else {
+				if (tile == docking_location) {
+					Axis axis_tc = DiagDirToAxis(rot);
+					if (axis != axis_tc) {
+						return_cmd_error(STR_ERROR_TOO_CLOSE_TO_ANOTHER_DOCK);
+					}
+				}
+			}
+		}
+
+		if (axis != AXIS_X && DiagDirToAxis(rot) == AXIS_X) {
+			if (IsTileType(tc + diff, MP_INDUSTRY) && GetIndustryGfx(tc + diff) == GFX_OILRIG_1 &&
```

Comment:
Will this work properly with NewGRF industries built on water with oil rig-style stations? (Such as fishing grounds in FIRS.)

Reply:
Probably yes.

I copied from these rules:
https://github.com/OpenTTD/OpenTTD/blob/b3b89257f745481db211fbc0bb185660528ec00c/src/industry_cmd.cpp#L737-L750


If FIRS is following these rules to create a station at the fishing grounds, then the docking tile should be in the same location as that of the oil rig.

Reply:
Yeah, checked, FIRS does use an "Oil Rig" tile which is exactly the `GFX_OILRIG_1` tile ID for both Fishing grounds and Dredring site industries. It should work then.